src/PathMappingRepository.php

```diff
@@ -0,0 +1,547 @@
+            'The passed resource must be a FilesystemResource or a ResourceCollection. Got: %s',
+            is_object($resource) ? get_class($resource) : gettype($resource)
+        ));
+    }
+
+    /**
+     * Removes all resources matching the given query.
+     *
+     * This method is not supported by this repository.
+     *
+     * @param string $query    A resource query.
+     * @param string $language The language of the query. All implementations
+     *                         must support the language "glob".
+     *
+     * @return integer The number of resources removed from the repository.
+     *
+     * @throws BadMethodCallException
+     */
+    public function remove($query, $language = 'glob')
+    {
+        throw new BadMethodCallException(sprintf('%s\(\) is not supported in %s', __METHOD__, __CLASS__));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function clear()
+    {
+        // Subtract root
+        $removed = $this->countStore() - 1;
+
+        $this->store->clear();
+        $this->createRoot();
+
+        return $removed;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function listChildren($path)
+    {
+        return new ArrayResourceCollection($this->getChildren($this->get($path)));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function hasChildren($path)
+    {
+        return $this->getChildren($this->get($path))->count() !== 0;
+    }
+
+
+    /**
+     * Find a resource by its path.
+     *
+     * @param string $path The path to resolve.
+     *
+     * @return \Puli\Repository\Api\Resource\Resource|null  The resource or null if
+     *                                                      the resource is not found.
+     */
+    private function resolveResource($path)
+    {
+        /*
+         * If the path exists in the store, return it directly
+         */
+        if ($this->store->exists($path)) {
+            $resolved = $this->createResource($this->store->get($path));
+            $resolved->attachTo($this, $path);
+
+            return $resolved;
+        }
+
+        /*
+         * Otherwise, we need to "resolve" it in two steps:
+         *      1.  find the resources from the store that are potential parent
+         *          of the path (we find them using regular expressions)
+         *      2.  for each of these potential parent, we try to find a real
+         *          file or directory on the filesystem and if we do find one,
+         *          we stop
+         */
+        $storePaths = array_reverse($this->store->keys());
+
+        // Create regex
+        $regExpressions = array();
+
+        foreach ($storePaths as $storePath) {
+            $prefix = rtrim($storePath, '/') . '/';
+            $regExpressions[] = '~^' . preg_quote($prefix, '~') . '~';
+        }
+
+        // Find potential parents
+        $resolved = null;
+
+        foreach ($regExpressions as $key => $regExpression) {
```

Comment:
> Currently, if multiple mappings are base paths for the searched path and the path is not found
> in the first mapping, this method fails even if it exists in the second mapping.

The `createResource()` and `$resource instanceof FilesystemResource` part is all about finding if the real filesystem file exists. It will stop the loop only if the filesystem resource does exists.

But you are right, I'll write a test for this.